### PR TITLE
feat: Improve installation robustness

### PR DESCRIPTION
**Pull request by TheAabedKhan**
Some screenshots
errors

### CODE REVIEW

File: assets/i18n/en_US.json [hunk 1]
```diff
@@ -163,13 +163,15 @@
     },
     "installerView": {
         "widgetTitle": "Installer",
-        "installType": "Select install type",
-        "installTypeDescription": "Select the installation type to proceed with.",
+        "installType": "Choose installation method",
+        "installTypeDescription": "To install, you can either mount or install normally. Mounting will install the patched app on top of an existing installation.",
```
Comment (Ushie) [inline assets/i18n/en_US.json:167]: What's the motivation behind the update of the description here? I believe the existing description is better fitting, the install methods can speak for themselves already

File: assets/i18n/en_US.json [hunk 2]
```diff
@@ -163,13 +163,15 @@
...
 
         "installButton": "Install",
         "installRootType": "Mount",
         "installNonRootType": "Normal",
 
+        "warning": "Disable auto updates after installing the app to avoid unexpected issues.",
```
Comment (Ushie) [inline assets/i18n/en_US.json:173]: Is this in scope of this PR?

File: assets/i18n/en_US.json [hunk 3]
```diff
@@ -327,5 +329,47 @@
         "integrationsContributors": "Integrations contributors",
         "cliContributors": "CLI contributors",
         "managerContributors": "Manager contributors"
+    },
+    "installErrorDialog": {
+        "title0": "Version mismatch",
+        "title1": "No installation found",
+        "title2": "Installation blocked",
+        "title3_0": "Installation blocked",
```
Comment (Ushie) [inline assets/i18n/en_US.json:337]: Duplicated strings

File: assets/i18n/en_US.json [hunk 4]
```diff
@@ -327,5 +329,47 @@
...
+
+        "callback5": "An existing installation of the app prevents the installation.",
+        "callback6": "The app could not be installed due to insufficient storage.",
+        "callback7": "The app is incompatible with this device.",
+        "callback8": "The installation took too long.",
+        "callback10": "To install by mounting the patched app, root access is required.",
+
+        "solution0": "Install the version of the app you are mounting and try again.",
+        "solution2": "Adjust your security settings and try again.",
+        "solution3_0": "Adjust your security settings and try again.",
```
Comment (Ushie) [inline assets/i18n/en_US.json:364]: Duplicated strings

File: assets/i18n/en_US.json [hunk 5]
```diff
@@ -327,5 +329,47 @@
...
+
+        "solution4": "Reinstall by uninstalling the app and try again?",
+        "solution4_0": "Uninstall the current version and try again?",
+
+        "solution5": "Uninstall the app and try again?",
```
Comment (Ushie) [inline assets/i18n/en_US.json:369]: Essentially the same in meaning, duplicated strings

File: assets/i18n/en_US.json [hunk 6]
```diff
@@ -327,5 +329,47 @@
...
+        "solution6": "Free up some space and try again.",
+        "solution7": "Contact the developer of the app and ask for support.",
+        "solution8": "Try again.",
```
Comment (Ushie) [inline assets/i18n/en_US.json:372]: This description feels underwhelming

**Review by Ushie — CHANGES_REQUESTED** [review]
I see multiple changes regarding the migration root install code, is that in scope for this PR? if so, I don't believe I'm fit to review it

File: assets/i18n/en_US.json (hunk 4 above)
Comment (oSumAtrIX) [reply assets/i18n/en_US.json:364]: Since these are two different solutions we treat both strings as two different solutions.

File: assets/i18n/en_US.json (hunk 6 above)
Comment (Ushie) [reply assets/i18n/en_US.json:372]: I would have to know what the fail code for 8 is, the string names don't help

File: assets/i18n/en_US.json (hunk 5 above)
Comment (oSumAtrIX) [reply assets/i18n/en_US.json:369]: solution4 can have the same string as solution5

**Review by Ushie — CHANGES_REQUESTED** [review]
Please name the strings appropiately, don't simplify them to "solutionx", it should be clear which fail code they target

File: assets/i18n/en_US.json (hunk 2 above)
Comment (oSumAtrIX) [reply assets/i18n/en_US.json:173]: Yes, it is necessary to let the user know what precautions to take after the installation

File: assets/i18n/en_US.json (hunk 5 above)
Comment (Ushie) [reply assets/i18n/en_US.json:369]: What is the difference between solution4 and solution4_0, and what prevents us from making common shared strings? it reduces repetition and unnecessary translation work, and the values of these strings are inconsistent

File: assets/i18n/en_US.json (hunk 1 above)
Comment (Ushie) [reply assets/i18n/en_US.json:167]: Perhaps the "overline" line in ListItems should be used in that case, that would allow all installation methods to have their own miniature description without relying on the dialog's description itself to describe it appropiately both in meaning and sentence structuring

**Comment by TheAabedKhan** [comment]
The X is the status code that's received from the PackageInstaller API. You can check the titleX to know what the solutionX is for.

File: assets/i18n/en_US.json (hunk 5 above)
Comment (oSumAtrIX) [reply assets/i18n/en_US.json:369]: Creating a common shared string means convoluting the current code. You'd use solution4 for status code 5 and without knowing whats going on this is confusing. Just because the strings are the same, doesn't mean they should be consolidated.

File: assets/i18n/en_US.json (hunk 6 above)
Comment (oSumAtrIX) [reply assets/i18n/en_US.json:372]: https://developer.android.com/reference/android/content/pm/PackageInstaller#STATUS_FAILURE_TIMEOUT

Comment (oSumAtrIX) [reply assets/i18n/en_US.json:372]: @TheAabedKhan Please add an enum and map the integers to them.

File: assets/i18n/en_US.json (hunk 5 above)
Comment (Ushie) [reply assets/i18n/en_US.json:369]: > You'd use solution4 for status code 5 and without knowing whats going on this is confusing

No, you'd use solution "reinstall_app"

Comment (oSumAtrIX) [reply assets/i18n/en_US.json:369]: Crowdin deduplicates strings, also the number suffix was used to prevent mapping the strings. If you want to change that, you'll need an enum as described in my other comment in a similar fashion I did in the compose PR.

File: assets/i18n/en_US.json (hunk 1 above)
Comment (Ushie) [reply assets/i18n/en_US.json:167]: image

The design in Figma can be used, for install methods that benefit from an explanation would use a trailing help icon button

File: assets/i18n/en_US.json (hunk 5 above)
Comment (Ushie) [reply assets/i18n/en_US.json:369]: The enum solution is good.

File: assets/i18n/en_US.json (hunk 1 above)
Comment (oSumAtrIX) [reply assets/i18n/en_US.json:167]: This can be allocated into a separate PR as it should not delay the current one rmore.

**Comment by Ushie** [comment]
Regarding the PR build, please see this: https://github.com/ReVanced/revanced-manager/issues/1540

**Review by Ushie — CHANGES_REQUESTED** [review]

File: assets/i18n/en_US.json [hunk 7]
```diff
@@ -327,5 +329,34 @@
...
         "cliContributors": "CLI contributors",
         "managerContributors": "Manager contributors"
+    },
+    "installErrorDialog": {
+        "mount_version_mismatch": "Version mismatch",
+        "mount_no_root": "No root access",
+        "mount_missing_installation": "No installation found",
+
+        "status_failure_blocked": "Installation blocked",
+        "install_failed_verification_failure": "Installation blocked",
```
[the remark this answers is not in the document]
Comment (TheAabedKhan) [reply assets/i18n/en_US.json:339]: @oSumAtrIX, for `install_failed_verification_failure`, maybe we should use `Verification failed`

[the remark this answers is not in the document]
Comment (oSumAtrIX) [reply assets/i18n/en_US.json:339]: Sureyvbut it is less clear for the user what it means

File: assets/i18n/en_US.json [hunk 8]
```diff
@@ -327,5 +329,34 @@
...
+
+        "status_failure_blocked": "Installation blocked",
+        "install_failed_verification_failure": "Verification failed",
+        "status_failure_invalid": "Installation invalid",
+        "install_failed_version_downgrade": "Can't downgrade",
+        "status_failure_conflict": "Installation conflict",
+        "status_failure_storage": "Installation storage issue",
+        "status_failure_incompatible": "Installation incompatible",
+        "status_failure_timeout": "Installation timeout",
+        "unknown_status": "Unknown status",
```
Comment (Ushie) [inline assets/i18n/en_US.json:346]:
```suggestion
        "status_unknown": "Installation failed due to an unknown reason",
```

**Review by Ushie — CHANGES_REQUESTED** [review]
The string key should be changed to status_unknown and maybe the string to "Installation failed due to an unknown reason"

File: lib/services/patcher_api.dart [hunk 9]
```diff
@@ -194,133 +202,305 @@ class PatcherAPI {
...
-  log.writeAsStringSync(logs);
-  FlutterFileDialog.saveFile(
-     params:SaveFileDialogParams(
-      sourceFilePath: log.path,
-      fileName: fileName,
-    ),
-  );
+  mountVersionMismatch(0),
+  mountMissingInstallation(1),
+  mountNoRoot(10);
```
Comment (Ushie) [inline lib/services/patcher_api.dart:465]: What are the numbers here? maybe they should be placed in order of the number?

**Review by Ushie — CHANGES_REQUESTED** [review]

File: lib/ui/views/installer/installer_viewmodel.dart [hunk 10]
```diff
@@ -388,15 +430,18 @@ class InstallerViewModel extends BaseViewModel {
   Future<void> installResult(BuildContext context, bool installAsRoot) async {
     try {
       _app.isRooted = installAsRoot;
```
[the remark this answers is not in the document]
Comment (TheAabedKhan) [reply lib/ui/views/installer/installer_viewmodel.dart:432]: Why remove it? This allows users to know whether the app was installed as root or non-root method from the App info section.

File: assets/i18n/en_US.json [hunk 11]
```diff
@@ -327,5 +329,34 @@
...
+        "status_failure_invalid": "Installation invalid",
+        "install_failed_version_downgrade": "Can't downgrade",
+        "status_failure_conflict": "Installation conflict",
+        "status_failure_storage": "Installation storage issue",
+        "status_failure_incompatible": "Installation incompatible",
+        "status_failure_timeout": "Installation timeout",
+        "status_unknown": "Installation failed",
+
+        "mount_version_mismatch_description": "To install by mounting the patched app, the version of the installed app must match the version of the patched app.\n\nInstall the version of the app you are mounting and try again.",
+        "mount_no_root_description": "To install by mounting the patched app, root access is required.\n\nGrant root access to ReVanced Manager and try again.",
```
Comment (Ushie) [inline assets/i18n/en_US.json:349]:
```suggestion
        "mount_no_root_description": "To install by mounting the patched app, root access must be granted.\n\nGrant root access to ReVanced Manager and try again.",
```

more consistent with the other strings

**Review by Ushie — CHANGES_REQUESTED** [review]
Changes otherwise look good, I currently don't have time to test so please test thoroughly before merging

File: assets/i18n/en_US.json [hunk 12]
```diff
@@ -327,5 +329,34 @@
...
+        "status_failure_storage": "Installation storage issue",
+        "status_failure_incompatible": "Installation incompatible",
+        "status_failure_timeout": "Installation timeout",
+        "status_unknown": "Installation failed",
+
+        "mount_version_mismatch_description": "To install by mounting the patched app, the version of the installed app must match the version of the patched app.\n\nInstall the version of the app you are mounting and try again.",
+        "mount_no_root_description": "To install by mounting the patched app, root access must be granted.\n\nGrant root access to ReVanced Manager and try again.",
+        "mount_missing_installation_description": "To install by mounting the patched app, the unpatched app must be installed on this device.\n\nInstall the app and try again.",
+
+        "status_failure_timeout_description": "Try again",
```
Comment (validcube) [inline assets/i18n/en_US.json:352]:
```suggestion
        "status_failure_timeout_description": "Try again.",
```

Comment (validcube) [reply assets/i18n/en_US.json:352]: The description feels too short to convey anything.

image

Something like this could work

```suggestion
        "status_failure_timeout_description": "The installation took too long and didn’t finish within the specified period.\n\nWould you like to try again?",
```

image

**Review by validcube — APPROVED** [review]

File: assets/i18n/en_US.json (hunk 12 above)
Comment (Ushie) [reply assets/i18n/en_US.json:352]: I agree, what would the specified period be to the user though? maybe best to reword that part

**Review by Ushie — CHANGES_REQUESTED** [review]
I have some string consistency changes locally that I forgot to push, I'll push it today once I get back home

File: lib/services/root_api.dart [hunk 13]
```diff
@@ -156,26 +150,22 @@ class RootAPI {
...
-    final String scriptFilePath = '$_postFsDataDirPath/$packageName.sh';
+    until [ "\$(getprop sys.boot_completed)" = 1 ]; do sleep 3; done
+    until [ -d "/sdcard/Android" ]; do sleep 1; done
+    
+    base_path=$_revancedDirPath/$packageName/base.apk
+    stock_path=\$(pm path $packageName | grep base | sed 's/package://g' )
+
+    chcon u:object_r:apk_data_file:s0  \$base_path
+    mount -o bind \$MIRROR\$base_path \$stock_path
+    '''
```
Comment (oSumAtrIX) [inline lib/services/root_api.dart:166]:
```suggestion

    am force-stop $packageName
    '''
```

**Review by Ushie — APPROVED** [review]
Code looks good, but don't have time to test, make sure to test before merging

File: assets/i18n/en_US.json (hunk 12 above)
Comment (validcube) [reply assets/i18n/en_US.json:352]: Hmm, how about:
*  "The installation took too long and didn’t finish within the specified period set by Android. Would you like to try again?"
* "The installation took too long and didn’t finish within the specified period set by your phone. Would you like to try again?"
* "The installation took too long and didn’t finish within the specified period set by your OEM. Would you like to try again?"

Comment (Ushie) [reply assets/i18n/en_US.json:352]: Maybe remove the specified period thing in its entirety, it just took too long, too long defined by who doesn't matter

Comment (validcube) [reply assets/i18n/en_US.json:352]: > Maybe remove the specified period thing in its entirety, it just took too long, too long defined by who doesn't matter

```suggestion
        "status_failure_timeout_description": "The installation took too long to finish.\n\nWould you like to try again?",
```

Feel a little too short in my opinion.
